src/Workspaces/Core/Portable/Utilities/ReferenceHolder`1.cs

```diff
@@ -0,0 +1,122 @@
+
+        public static ReferenceHolder<T> Weak(T value)
+        {
+            if (value is null)
+            {
+                // Track this as a strong reference so we know 'Equals' should only look at values originally null.
+                return Strong(value);
+            }
+
+            return new ReferenceHolder<T>(new WeakReference<T>(value), RuntimeHelpers.GetHashCode(value));
+        }
+
+        [return: MaybeNull]
+        public T TryGetTarget()
+        {
+            if (_weakReference is object)
+                return _weakReference.GetTarget();
+
+            return _strongReference;
+        }
+
+        public override bool Equals(object? obj)
+        {
+            return obj is ReferenceHolder<T> other
+                && Equals(other);
+        }
+
+        public bool Equals(ReferenceHolder<T> other)
+        {
+            var x = TryGetTarget();
+            var y = other.TryGetTarget();
+            if (x is null)
+            {
+                if (_weakReference is object)
+                {
+                    // 'x' is a weak reference that was collected. Verify 'y' is a collected weak reference with the
+                    // same runtime hash code.
+                    return y is null && other._weakReference is object && _hashCode == other._hashCode;
+                }
+                else
+                {
+                    // Null values are equal iff both were originally references to null.
+                    return y is null && other._weakReference is null;
+                }
+            }
+
+            // Intentional reference equality check
+            return x == y;
+        }
+
+        public override int GetHashCode()
+        {
+            if (_weakReference is object)
+                return _hashCode;
+
+            // RuntimeHelpers.GetHashCode allows null arguments
+            return RuntimeHelpers.GetHashCode(_strongReference!);
+        }
+
+        internal TestAccessor GetTestAccessor()
+        {
+            return new TestAccessor(this);
+        }
+
+        internal readonly struct TestAccessor
+        {
+#pragma warning disable IDE0052 // Remove unread private members
+            private readonly ReferenceHolder<T> _referenceHolder;
+#pragma warning restore IDE0052 // Remove unread private members
+
+            internal TestAccessor(ReferenceHolder<T> referenceHolder)
+            {
+                _referenceHolder = referenceHolder;
+            }
+
+            /// <summary>
+            /// Creates a <see cref="ReferenceHolder{T}"/> for a weakly-held reference that has since been collected.
+            /// </summary>
+            /// <param name="hashCode">The hash code of the collected value.</param>
+            /// <returns>A weak <see cref="ReferenceHolder{T}"/> which was already collected.</returns>
+            public static ReferenceHolder<T> ReleasedWeak(int hashCode)
+                => new ReferenceHolder<T>(new WeakReference<T>(null!), hashCode);
```

Comment:
You may be able to leverage https://github.com/dotnet/roslyn/blob/master/src/Test/Utilities/Portable/ObjectReference.cs to avoid having to have a test accessor at all; that way you can test the actual weak behavior versus this mock-up of it. That said this is at least fairly easy to understand, so it's good.

Reply:
ObjectReference is not deterministic and is slow, so I opted for a better performing and deterministic alternative.